misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJobConsumer.kt

```diff
@@ -63,73 +69,116 @@ internal class SqsJobConsumer @Inject internal constructor(
   }
 
   internal inner class QueueReceiver(
-    private val queueName: QueueName,
+    queueName: QueueName,
     private val handler: JobHandler
   ) {
     private val queue = queues[queueName]
 
-    fun runOnce(): Status {
-      val (receiveDuration, messages) = queue.call { client ->
-        val receiveRequest = ReceiveMessageRequest()
-          .withAttributeNames("All")
-          .withMessageAttributeNames("All")
-          .withQueueUrl(queue.url)
-          .withMaxNumberOfMessages(config.message_batch_size)
+    fun run(): Status {
+      // Receive messages in parallel. Default to 1 if feature flag is not defined.
+      val numReceivers = receiversForQueue()
+      val futures = (1..numReceivers)
+          .filter { num ->
+            val lease = leaseManager.requestLease("sqs-job-consumer-${queue.name.value}-$num")
+            !config.clustered_consumers || lease.checkHeld()
+          }
+          .map {
+            CompletableFuture.supplyAsync(Supplier {
+              receive()
+            }, receivingThreads)
+          }
 
-        timed { client.receiveMessage(receiveRequest).messages }
+      // Either all messages are consumed and processed successfully, or we signal failure.
+      // If none of the received consume any messages, return NO_WORK for backoff.
+      return CompletableFuture.allOf(*futures.toTypedArray()).thenApply {
+        futures.flatMap { it.join() }
+            .fold(Status.NO_WORK) { finalStatus, status ->
+              when {
+                status == Status.FAILED -> status
+                status == Status.OK && finalStatus != Status.FAILED -> status
+                status == Status.NO_WORK && finalStatus == Status.NO_WORK -> status
+                status !in listOf(Status.OK, Status.NO_WORK, Status.FAILED) ->
```

Comment:
shouldn't the else just be throw? why do you check if it's in the set of all values?

Reply:
This algorithm knows how to handle `Status.OK, Status.NO_WORK, Status.FAILED`. I could add this check outside the `fold` if this makes this more clear. 
The `else` is maintaining the accumulator value if none of the previous conditions are met. For example, for the list [Status.FAILED, Status.OK], I want to make sure I return `Status.FAILED`

Reply:
got it. there is another status `NO_RESCHEDULE`

Reply:
yeah, this code is not using that status. I'm trying to protect if someone added it later, but misses this bit.